src/posix/platform/resolver.cpp

```diff
@@ -129,15 +144,109 @@ void Resolver::LoadDnsServerListFromConf(void)
     return;
 }
 
-void Resolver::Query(otPlatDnsUpstreamQuery *aTxn, const otMessage *aQuery)
+#if OPENTHREAD_CONFIG_BORDER_ROUTING_ENABLE
+void Resolver::BorderRoutingRdnssCallback(void *aResolver)
+{
+    static_cast<Resolver *>(aResolver)->BorderRoutingRdnssCallback();
+}
+
+void Resolver::BorderRoutingRdnssCallback(void)
 {
-    char         packet[kMaxDnsMessageSize];
-    otError      error  = OT_ERROR_NONE;
-    uint16_t     length = otMessageGetLength(aQuery);
+    otBorderRoutingPrefixTableIterator iterator;
+    otBorderRoutingRdnssAddrEntry      entry;
+    otBorderRoutingRdnssAddrEntry      rdnssEntries[kMaxRecursiveServerCount + 1];
+    otIp6Address                       rdnssServers[kMaxRecursiveServerCount];
+    uint32_t                           numEntries = 0;
+
+    otBorderRoutingPrefixTableInitIterator(gInstance, &iterator);
+
+    while (otBorderRoutingGetNextRdnssAddrEntry(gInstance, &iterator, &entry) == OT_ERROR_NONE)
+    {
+        uint32_t i = 0;
+
+        // Check if the entry address is already in the list.
+        for (; i < numEntries; ++i)
+        {
+            if (otIp6IsAddressEqual(&entry.mAddress, &rdnssEntries[i].mAddress))
+            {
+                rdnssEntries[i].mLifetime = OT_MAX(rdnssEntries[i].mLifetime, entry.mLifetime);
+
+                break;
+            }
+        }
+
+        // If the address is not a duplicate, add the entry to the entry list.
+        if (i == numEntries)
+        {
+            rdnssEntries[numEntries++] = entry;
+
+            std::sort(rdnssEntries, rdnssEntries + numEntries,
+                      [](const otBorderRoutingRdnssAddrEntry &a, const otBorderRoutingRdnssAddrEntry &b) {
+                          if (a.mLifetime != b.mLifetime)
+                          {
+                              return a.mLifetime > b.mLifetime;
+                          }
+                          else
+                          {
+                              // If lifetimes are equal, prefer the one with the larger numeric values
+                              return memcmp(&a, &b, sizeof(otIp6Address)) > 0;
```

Comment:
This is not correct and wont compare the addresses.
Note that `a` and `b` are `otBorderRoutingRdnssAddrEntry` struct which is
```c++
typedef struct otBorderRoutingRdnssAddrEntry
{
    otBorderRoutingRouterEntry mRouter;              ///< Information about the router advertising this address.
    otIp6Address               mAddress;             ///< The DNS Server IPv6 address.
    uint32_t                   mMsecSinceLastUpdate; ///< Milliseconds since last update of this address.
    uint32_t                   mLifetime;            ///< Lifetime of the address (in seconds).
} otBorderRoutingRdnssAddrEntry;
```

Therefore `&a` refers to its first field `otBorderRoutingRouterEntry mRouter` and not its `mAddress`.

Should use `&a.mAddress`  instead.

Reply:
We generally avoid `memcmp` and `memcpy` and/or try to be extra careful when using them, as they remove any type checking (they treat everything as raw `void *` and ignore whatever the passed-in input type is).